### PR TITLE
[WIP] Use MSB radix sort in sortBlock and ColumnVector

### PR DESCRIPTION
**Pull request by kvinty**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- Performance Improvement

Short description (up to few sentences):
Use partial MSB radix sort instead of std::partial_sort

### CODE REVIEW

**Comment by kvinty** [comment]
Run performance tests
Need to fix floats comparison with nan_direction_hint

File: dbms/src/Interpreters/sortBlock.cpp [hunk 1]
```diff
@@ -149,7 +150,7 @@ void sortBlock(Block & block, const SortDescription & description, UInt64 limit)
             PartialSortingLessWithCollation less_with_collation(columns_with_sort_desc);
 
             if (limit)
-                std::partial_sort(perm.begin(), perm.begin() + limit, perm.end(), less_with_collation);
+                RadixSort<RadixSortNumTraits<size_t>, PartialSortingLessWithCollation>::executeMSD(perm.data(), perm.size(), limit, less_with_collation);
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/sortBlock.cpp]: You're sorting permutation by indices not by values. How it can work at all?

**Comment by alexey-milovidov** [comment]
?

File: dbms/src/Common/RadixSort.h [hunk 2]
```diff
@@ -222,7 +209,7 @@ struct RadixSort
      * Puts elements to buckets based on PASS-th digit, then recursively calls insertion sort or itself on the buckets
      */
     template <size_t PASS>
-    static inline void radixSortMSDInternal(Element * arr, size_t size, size_t limit)
+    static inline void radixSortMSDInternal(Element * arr, size_t size, size_t limit, Less less __attribute__((unused)))
```
Comment (alexey-milovidov) [inline dbms/src/Common/RadixSort.h]: There are C++17 attributes.

File: dbms/src/Common/RadixSort.h [hunk 3]
```diff
@@ -172,8 +154,13 @@ using RadixSortNumTraits =
...
         RadixSortFloatTraits<T>>;
 
+template <typename T>
+struct DefaultLess
+{
+    bool less(const T &a, const T &b) { return a < b; }
+};
 
-template <typename Traits>
+template <typename Traits, typename Less=DefaultLess<typename Traits::Key>>
```
Comment (alexey-milovidov) [inline dbms/src/Common/RadixSort.h]: Whitespaces.

File: dbms/src/Common/RadixSort.h [hunk 4]
```diff
@@ -172,8 +154,13 @@ using RadixSortNumTraits =
             RadixSortIntTraits<T>>,
         RadixSortFloatTraits<T>>;
 
+template <typename T>
+struct DefaultLess
+{
+    bool less(const T &a, const T &b) { return a < b; }
```
Comment (alexey-milovidov) [inline dbms/src/Common/RadixSort.h]: std::less.

File: dbms/src/Columns/ColumnVector.cpp [hunk 5]
```diff
@@ -104,9 +104,9 @@ void ColumnVector<T>::getPermutation(bool reverse, size_t limit, int nan_directi
             res[i] = i;
 
         if (reverse)
-            std::partial_sort(res.begin(), res.begin() + limit, res.end(), greater(*this, nan_direction_hint));
+            RadixSort<RadixSortNumTraits<size_t>, greater>::executeMSD(res.data(), res.size(), limit, greater(*this, nan_direction_hint));
```
Comment (alexey-milovidov) [inline dbms/src/Columns/ColumnVector.cpp]: You still sort by index, not by value.

**Comment by alexey-milovidov** [comment]
> No radix sort in sortBlock

Why?

**Comment by alexey-milovidov** [comment]
?

**Comment by alexey-milovidov** [comment]
It's abandoned, but will be continued in a different form.

**Pull request closed without merging**